File: src/Client/Private/API/MikanStencilAPI.cpp
#include "MikanStencilAPI.h"
#include "MikanRequestManager.h"
#include "MikanStencilTypes_json.h"
#include "MikanStencilTypes_binary.h"

MikanStencilAPI::MikanStencilAPI(class MikanRequestManager* requestManager)
	: m_requestManager(requestManager)
{
	m_requestManager->addTextResponseFactory<MikanStencilList>();
	m_requestManager->addTextResponseFactory<MikanStencilQuadInfo>();
	m_requestManager->addTextResponseFactory<MikanStencilBoxInfo>();
	m_requestManager->addTextResponseFactory<MikanStencilModelInfo>();
	m_requestManager->addBinaryResponseFactory<MikanStencilModelRenderGeometry>();
}

MikanResponseFuture MikanStencilAPI::getQuadStencilList() // returns MikanStencilList
{
	return m_requestManager->sendRequest(k_getQuadStencilList);
}

MikanResponseFuture MikanStencilAPI::getQuadStencil(MikanStencilID stencilId) // returns MikanStencilQuad
{
	return m_requestManager->sendRequestWithPayload<int>(k_getQuadStencil, stencilId);
}

MikanResponseFuture MikanStencilAPI::getBoxStencilList() // returns MikanStencilList
{
	return m_requestManager->sendRequest(k_getBoxStencilList);
}

MikanResponseFuture MikanStencilAPI::getBoxStencil(MikanStencilID stencilId) // returns MikanStencilBox
{
	return m_requestManager->sendRequestWithPayload<int>(k_getBoxStencil, stencilId);
}

MikanResponseFuture MikanStencilAPI::getModelStencilList() // returns MikanStencilList
{
	return m_requestManager->sendRequest(k_getModelStencilList);
}

MikanResponseFuture MikanStencilAPI::getModelStencil(MikanStencilID stencilId) // returns MikanStencilModel
{
	return m_requestManager->sendRequestWithPayload<int>(k_getModelStencil, stencilId);
}

MikanResponseFuture MikanStencilAPI::getModelStencilRenderGeometry(MikanStencilID stencilId) // returns MikanStencilModel
{
	return m_requestManager->sendRequestWithPayload<int>(k_getModelStencilRenderGeometry, stencilId);
}

File: bindings/csharp/MikanStencilAPI.cs
using System.Threading.Tasks;

namespace MikanXR
{	
	public class MikanStencilAPI
	{
		private MikanRequestManager _requestManager;

		public MikanStencilAPI(MikanRequestManager requestManager)
		{
			_requestManager= requestManager;
			_requestManager.AddTextResponseFactory<MikanStencilList>();
			_requestManager.AddTextResponseFactory<MikanStencilQuadInfo>();
			_requestManager.AddTextResponseFactory<MikanStencilBoxInfo>();
			_requestManager.AddTextResponseFactory<MikanStencilModelInfo>();
			_requestManager.AddBinaryResponseFactory<
				MikanStencilModeRenderGeometryFactory,
				MikanStencilModelRenderGeometry>();
		}

		private static readonly string k_getQuadStencilList = "getQuadStencilList";
		public Task<MikanResponse> getQuadStencilList() // returns MikanStencilList
		{
			return _requestManager.SendRequest(k_getQuadStencilList);
		}

		private static readonly string k_getQuadStencil = "getQuadStencil";
		public Task<MikanResponse> getQuadStencil(int stencilId) // returns MikanStencilQuad
		{
			return _requestManager.SendRequestWithPayload<int>(k_getQuadStencil, stencilId);
		}

		private static readonly string k_getBoxStencilList = "getBoxStencilList";
		public Task<MikanResponse> getBoxStencilList() // returns MikanStencilList
		{
			return _requestManager.SendRequest(k_getBoxStencilList);
		}

		private static readonly string k_getBoxStencil = "getBoxStencil";
		public Task<MikanResponse> getBoxStencil(int stencilId) // returns MikanStencilBox
		{
			return _requestManager.SendRequestWithPayload<int>(k_getBoxStencil, stencilId);
		}

		private static readonly string k_getModelStencilList = "getModelStencilList";
		public Task<MikanResponse> getModelStencilList() // returns MikanStencilList
		{
			return _requestManager.SendRequest(k_getModelStencilList);
		}

		private static readonly string k_getModelStencil = "getModelStencil";
		public Task<MikanResponse> getModelStencil(int stencilId) // returns MikanStencilModel
		{
			return _requestManager.SendRequestWithPayload<int>(k_getModelStencil, stencilId);
		}

		private static readonly string k_getModelStencilRenderGeometry = "getModelStencilRenderGeometry";
		public Task<MikanResponse> getModelStencilRenderGeometry(int stencilId) // returns MikanStencilModelMesh
		{
			return _requestManager.SendRequestWithPayload<int>(k_getModelStencilRenderGeometry, stencilId);
		}
	}
}